src/Agent.Plugins/BuildDrop/BuildDropPlugin.cs

```diff
@@ -0,0 +1,186 @@
+            }
+
+            // Project ID
+            Guid projectId = new Guid(context.Variables.GetValueOrDefault(BuildVariables.TeamProjectId)?.Value ?? Guid.Empty.ToString());
+            ArgUtil.NotEmpty(projectId, nameof(projectId));
+
+            // Build ID
+            string buildIdStr = context.Variables.GetValueOrDefault(BuildVariables.BuildId)?.Value ?? string.Empty;
+            if (!int.TryParse(buildIdStr, out int buildId))
+            {
+                // This should not happen since the build id comes from build environment. But a user may override that so we must be careful.
+                throw new ArgumentException(StringUtil.Loc("BuildIdIsNotValid", buildIdStr));
+            }
+
+            string fullPath = Path.GetFullPath(targetPath);
+            bool isFile = File.Exists(fullPath);
+            bool isDir = Directory.Exists(fullPath);
+            if (!isFile && !isDir)
+            {
+                // if local path is neither file nor folder
+                throw new FileNotFoundException(StringUtil.Loc("PathNotExist", targetPath));
+            }
+            else if (isDir && Directory.EnumerateFiles(fullPath, "*", SearchOption.AllDirectories).FirstOrDefault() == null)
+            {
+                // if local path is a folder which contains nothing
+                throw new ArgumentException(StringUtil.Loc("DirectoryIsEmptyForArtifact", fullPath, artifactName));
+            }
+
+            // Upload to VSTS BlobStore, and associate the artifact with the build.
+            context.Output(StringUtil.Loc("UploadingBuildDrop", fullPath, buildId));
+            BuildDropServer server = new BuildDropServer();
+            await server.UploadDropArtifactAsync(context, projectId, buildId, artifactName, fullPath, token);
+            context.Output(StringUtil.Loc("UploadArtifactFinished"));
+        }
+    }
+
+    // CAller: DownloadBuildDrop task
+    // Can be invoked from a build run or a release run should a build be set as the artifact. 
+    public class DownloadBuildDropTask : BuildDropTaskPluginBase
+    {
+        // Same as https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/DownloadBuildDropV0/task.json
+        public override Guid Id => new Guid("61F2A582-95AE-4948-B34D-A1B3C4F6A737");
+
+        protected override async Task ProcessCommandInternalAsync(
+            AgentTaskPluginExecutionContext context, 
+            string targetPath, 
+            string artifactName,
+            CancellationToken token)
+        {
+            // Create target directory if absent
+            string fullPath = Path.GetFullPath(targetPath);
+            bool isDir = Directory.Exists(fullPath);
+            if (!isDir)
+            {
+                Directory.CreateDirectory(fullPath);
+            }
+
+            // Project ID
+            Guid projectId = new Guid(context.Variables.GetValueOrDefault(BuildVariables.TeamProjectId)?.Value ?? Guid.Empty.ToString());
```

Comment:
I didn't write that myself (copied from existing code) but I believe that's to achieve a one-liner with ?? since that operator cannot be used between incompatible types such as string vs guid.

I replaced the whole line anyway with right constant.